kms/tpmkms/tpmkms.go

```diff
@@ -0,0 +1,674 @@
+func New(ctx context.Context, opts apiv1.Options) (kms *TPMKMS, err error) {
+	kms = &TPMKMS{}
+	tpmOpts := []tpm.NewTPMOption{tpm.WithStore(storage.BlackHole())} // TODO(hs): use some default storage location instead? Or in-memory implementation?
+	if opts.URI != "" {
+		u, err := uri.ParseWithScheme(Scheme, opts.URI)
+		if err != nil {
+			return nil, fmt.Errorf("failed parsing %q as URI: %w", opts.URI, err)
+		}
+		if device := u.Get("device"); device != "" {
+			tpmOpts = append(tpmOpts, tpm.WithDeviceName(device))
+		}
+		if storageDirectory := u.Get("storage-directory"); storageDirectory != "" {
+			tpmOpts = append(tpmOpts, tpm.WithStore(storage.NewDirstore(storageDirectory)))
+		}
+		kms.attestationCABaseURL = u.Get("attestation-ca-url")
+		kms.attestationCARootFile = u.Get("attestation-ca-root")
+		kms.attestationCAInsecure = u.GetBool("attestation-ca-insecure")
+		kms.permanentIdentifier = u.Get("permanent-identifier") // TODO(hs): determine if this is needed
+	}
+
+	kms.tpm, err = tpm.New(tpmOpts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed creating new TPM: %w", err)
+	}
+
+	return
+}
+
+// CreateKey generates a new key in the TPM KMS and returns the public key.
+//
+// The `name` in the [apiv1.CreateKeyRequest] can be used to specify
+// some key properties. These are as follows:
+//
+//   - name=<name>: specify the name to identify the key with
+//   - ak=true: if set to true, an Attestation Key (AK) will be created instead of an application key
+//   - attest-by=<akName>: attest an application key at creation time with the AK identified by `akName`
+//   - qualifying-data=<random>: hexadecimal coded binary data that can be used to guarantee freshness when attesting creation of a key
+//
+// Some examples usages:
+//
+// Create an application key, without attesting it:
+//
+//	tpmkms:name=my-key
+//
+// Create an Attestation Key (AK):
+//
+//	tpmkms:name=my-ak;ak=true
+//
+// Create an application key, attested by `my-ak` with "1234" as the Qualifying Data:
+//
+//	tpmkms:name=my-attested-key;attest-by=my-ak;qualifying-data=61626364
+func (k *TPMKMS) CreateKey(req *apiv1.CreateKeyRequest) (*apiv1.CreateKeyResponse, error) {
+	switch {
+	case req.Name == "":
+		return nil, errors.New("createKeyRequest 'name' cannot be empty")
+	case req.Bits < 0:
+		return nil, errors.New("createKeyRequest 'bits' cannot be negative")
+	}
+
+	properties, err := parseNameURI(req.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed parsing %q: %w", req.Name, err)
+	}
+
+	v, ok := signatureAlgorithmMapping[req.SignatureAlgorithm]
+	if !ok {
+		return nil, fmt.Errorf("TPMKMS does not support signature algorithm %q", req.SignatureAlgorithm)
+	}
+
+	if properties.ak && v.Type == "ECDSA" {
+		return nil, errors.New("AKs must be RSA keys")
+	}
+
+	if properties.ak && req.Bits != defaultRSAAKSize { // 2048
+		return nil, fmt.Errorf("creating %d bit AKs is not supported; AKs must be RSA 2048 bits", req.Bits)
+	}
```

Comment:
req.Bits == 0 should be valid if properties.ak is true.

Reply:
Fixed in dc0a70717a587b95079a181a630c83126f150c2b 😓